vcr/verifier/credential_status.go

```diff
@@ -0,0 +1,221 @@
+			// cannot happen. already validated in credential.defaultCredentialValidator{}
+			return err
+		}
+		revoked, err := sList.expanded.Bit(index)
+		if err != nil {
+			return err
+		}
+		if revoked {
+			return types.ErrRevoked
+		}
+	}
+	return nil
+}
+
+func (cs *credentialStatus) statusList(statusListCredential string) (*statusList, error) {
+	// TODO: check if there is a cached version to return
+	return cs.update(statusListCredential)
+}
+
+// update
+func (cs *credentialStatus) update(statusListCredential string) (*statusList, error) {
+	// download and verify
+	cred, err := cs.download(statusListCredential)
+	if err != nil {
+		return nil, err
+	}
+	credSubject, err := cs.verifyStatusList2021Credential(*cred)
+	if err != nil {
+		return nil, err
+	}
+
+	// make statusList
+	expanded, err := statuslist2021.Expand(credSubject.EncodedList)
+	if err != nil {
+		// cant happen, already checked in verifyStatusList2021Credential
+		return nil, err
+	}
+	sl := statusList{
+		credential:           cred,
+		statusListCredential: statusListCredential,
+		statusPurpose:        credSubject.StatusPurpose,
+		expanded:             expanded,
+		lastUpdated:          time.Now(),
+	}
+	// TODO: cache updated credential so it does not have to be downloaded everytime
+	//  	 also cache if statusPurposes != 'revocation' to prevent unnecessary downloads
+	return &sl, nil
+}
+
+// download the StatusList2021Credential found at statusList2021Entry.statusListCredential
+func (cs *credentialStatus) download(statusListCredential string) (*vc.VerifiableCredential, error) {
+	var cred vc.VerifiableCredential // VC containing CredentialStatus of the credentialToVerify
+	req, err := http.NewRequest(http.MethodGet, statusListCredential, new(bytes.Buffer))
+	if err != nil {
+		return nil, err
+	}
+	res, err := cs.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	body, err := io.ReadAll(res.Body)
+	if err = res.Body.Close(); err != nil {
+		// log, don't fail
+		log.Logger().
+			WithError(err).
+			WithField("method", "credentialStatus.download").
+			Debug("failed to close response body")
+	}
+	if res.StatusCode > 299 || err != nil {
+		return nil, fmt.Errorf("fetching StatusList2021Credential from '%s' failed", statusListCredential)
+	}
+	if err = json.Unmarshal(body, &cred); err != nil {
+		return nil, err
+	}
+	return &cred, nil
+}
+
+// verifyStatusList2021Credential checks that the StatusList2021Credential is currently valid
+func (cs *credentialStatus) verifyStatusList2021Credential(cred vc.VerifiableCredential) (*credential.StatusList2021CredentialSubject, error) {
+	// make sure we have the correct credential.
+	if len(cred.Type) > 2 || !cred.IsType(ssi.MustParseURI(credential.StatusList2021CredentialType)) {
+		return nil, errors.New("incorrect credential types")
+	}
+
+	// returns statusList2021CredentialValidator, or Validate() fails because base type is missing
+	if err := credential.FindValidator(cred).Validate(cred); err != nil {
```

Comment:
probably should just directly create the validator, since we exactly know which one we need...

Reply:
added to note on epic. dependency cycles vs consistency are a big issue in this feature